File: day3/solution.py
from pathlib import Path
import re

current_dir = Path(__file__).parent
input_file = Path(current_dir, "input.txt")

text = input_file.read_text()

def parse(text):
    instr = re.findall(r"(mul\(\d+\,\d+\)|do\(\)|don't\(\))", text)
    res = []
    for t in instr:
        op, rest = t.split("(")
        res.append((op, *[int(d) for d in rest[:-1].split(",") if d]),)

    return res

def part1():
    return sum(t[1] * t[2] for t in parse(text) if t[0] == "mul")

def part2():
    enabled, tot = True, 0

    for t in parse(text):
        if t[0] == "mul" and enabled:
            tot += t[1] * t[2]
        else:
            enabled = True if t[0] == "do" else False

    return tot

print(part1())
print(part2())

File: day3/solution.exs
defmodule Solution do
  def parse(text) do
    Regex.scan(~r/(mul\(\d{1,3}\,\d{1,3}\)|do\(\)|don't\(\))/i, text)
    |> Enum.map(&Enum.uniq/1)
    |> Enum.map(&hd/1)
    |> Enum.map(fn instr ->
      [op | args] = String.split(instr, ~r/[\(\),]/, trim: true)
      [op | Enum.map(args, &String.to_integer/1)]
    end)
  end

  def part1(instructions) do
    instructions
    |> Enum.filter(&match?(["mul" | _], &1))
    |> Enum.map(fn [_, a, b] -> a * b end)
    |> Enum.sum()
  end

  def part2(instructions) do
    instructions
    |> Enum.reduce({0, true}, &process_instruction/2)
    |> elem(0)
  end

  defp process_instruction(["mul", a, b], {tot, true}), do: {a * b + tot, true}
  defp process_instruction(["do" | _], {tot, _}), do: {tot, true}
  defp process_instruction(["don't" | _], {tot, _}), do: {tot, false}
  defp process_instruction(_, acc), do: acc
end

text = File.read!("./input.txt")

Solution.parse(text)
|> Solution.part1()
|> IO.inspect()

Solution.parse(text)
|> Solution.part2()
|> IO.inspect()
